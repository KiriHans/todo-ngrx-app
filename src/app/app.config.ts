import { ApplicationConfig, isDevMode } from '@angular/core';
import { provideRouter } from '@angular/router';

import { provideAnimations } from '@angular/platform-browser/animations';

import { routes } from './app.routes';
import { provideStore } from '@ngrx/store';
import { provideStoreDevtools } from '@ngrx/store-devtools';
import { appReducers, metaReducers } from './app.reducer';

export const appConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes),
    provideAnimations(),
    provideStore(appReducers, { metaReducers }),
    provideStoreDevtools({ maxAge: 25, logOnly: !isDevMode() }),
  ],
};
