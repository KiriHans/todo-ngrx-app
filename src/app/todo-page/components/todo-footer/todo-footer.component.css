:host{
  --elevated-card-color: rgb(241, 241, 241);
  --elevation-card: 0px 2px 3px -1px rgba(0, 0, 0, .2), 0px 1px 2px 0px rgba(0, 0, 0, .14), 0px 1px 3px 0px rgba(0, 0, 0, .12);
}

footer{
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
  flex-wrap: wrap;
  
  padding: 1rem;
  
  font-size: 0.8rem;

}

@media (width > 30rem) {
  footer{
    padding: 0.3rem 0.5rem;
    min-height: 2.6rem;
  }
  
}

.todo-count{
  font-weight: 400;
}

ul{
  display: flex;
  flex-direction: row;
  justify-content: center;
  gap: 0.5rem;

  padding: 0;
  margin: 0.2rem 0;
  
}

li{
  list-style: none;
  
}

button, a{
  font-size: 0.8rem;
  font-weight: 400;

}

.active-filter{
  border-color: var(--elevated-card-color);
  box-shadow: var(--elevation-card);
}