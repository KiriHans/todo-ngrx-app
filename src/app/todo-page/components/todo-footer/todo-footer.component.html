<footer>
  <span class="todo-count">
    <strong>{{ itemsLeftList().length }}</strong> items left
  </span>
  <ul class="filters" *ngIf="todoList().length !== 0">
    <li *ngFor="let filter of filterList">
      <a
        [routerLink]="'.'"
        [fragment]="filter"
        (click)="changeFilter(filter)"
        *ngIf="filter !== currentFilter()"
        mat-button
      >
        {{ filter | titlecase }}
      </a>
      <a
        [routerLink]="'.'"
        [fragment]="filter"
        *ngIf="filter === currentFilter()"
        mat-button
        routerLinkActive="active-filter"
      >
        {{ filter | titlecase }}
      </a>
    </li>
  </ul>
  <button
    class="clear-completed"
    [routerLink]="'.'"
    fragment="all"
    (click)="clearCompletedTodos()"
    mat-flat-button
    *ngIf="todoList().length !== 0"
  >
    Clear completed
  </button>
</footer>
