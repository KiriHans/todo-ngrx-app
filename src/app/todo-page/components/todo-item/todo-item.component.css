:host{
	width: 100%;
}

:host ::ng-deep :is(.field-edit-todo) .mat-mdc-form-field-subscript-wrapper {
	display: none;
}

.edit{
  display: none;
}


li{
	list-style: none;
	display: flex;
	justify-content: space-between;
	flex-direction: row;
	align-items: center;
}

.view{
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	gap: 1rem;

	height: 3.2rem;

	width: 100%;
}

.view label{
	display: flex;
	flex-direction: row;
	align-items: center;

	width: 100%;
	height: 2rem;

	font-size: 1rem;

}

.view mat-checkbox {
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	transform: scale(1.2);
}


li.editing .view {
	display: none;
}


li.editing .edit {
	font-family: roboto;
	font-size: 1rem;
	display: block;
	width: calc(100% - 80px);
	height: 3rem;
	padding:  0 16px;
	margin: 0 0 0 40px;

	border: 2px solid #ff4081;

}

.edit:focus{
	outline: none;
	box-shadow: 0 0 1px 1px rgba(255, 64, 127, .5);
	
}

.completed{
	color: #949494;
	text-decoration: line-through;
}

.completed .edit {
	color: #949494;
	text-decoration: line-through;

	border: 2px solid hsla(340, 100%, 63%, 0.732);
	
}

.completed .destroy{
	color: #ff4081;
	font-variation-settings:
  'FILL' 0,
  'wght' 600,
  'GRAD' 0,
  'opsz' 48
}