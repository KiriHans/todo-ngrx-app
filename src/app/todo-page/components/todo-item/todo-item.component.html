<li [ngClass]="{ completed: todo.completed, editing: editing }">
  <div class="view">
    <mat-checkbox [formControl]="completed"> </mat-checkbox>

    <label (dblclick)="editTodo()" for="add">
      {{ todo.text }}
    </label>
    <button (click)="deleteTodo()" mat-icon-button class="destroy">
      <mat-icon>close</mat-icon>
    </button>
  </div>

  <input
    [cdkTrapFocusAutoCapture]="editing"
    [cdkTrapFocus]="editing"
    [formControl]="inputTodo"
    [value]="todo.text"
    (blur)="stopEdit()"
    (keyup.enter)="stopEdit()"
    appSelect
    class="edit"
    type="text"
  />
</li>
