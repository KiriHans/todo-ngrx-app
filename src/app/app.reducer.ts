import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import { TodoItem } from './todo-page/models/todo-item.model';
import { todoReducer } from './todo-page/store/reducers/todo.reducer';
import { filterReducer } from './todo-page/store/reducers/filter.reducer';
import { ValidFilters } from './todo-page/types/valid-filter.type';
import { storageMetaReducer } from './todo-page/store/reducers/todo.metareducer';

export interface AppState {
  todos: TodoItem[];
  filter: ValidFilters;
}

export const appReducers: ActionReducerMap<AppState> = {
  todos: todoReducer,
  filter: filterReducer,
};

export const metaReducers: MetaReducer[] = [storageMetaReducer];
