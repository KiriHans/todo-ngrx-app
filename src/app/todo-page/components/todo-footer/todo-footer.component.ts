import { Component, OnInit, Signal } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatButtonModule } from '@angular/material/button';
import { Store } from '@ngrx/store';
import { selectFilter } from '../../store/selectors/filter.selectors';
import { ActivatedRoute, RouterModule } from '@angular/router';

import * as FilterActions from '../../store/actions/filter.actions';
import * as TodoActions from '../../store/actions/todo.actions';

import { selectPendingTodos, selectTodoList } from '../../store/selectors/todo.selectors';
import { TodoItem } from '../../models/todo-item.model';
import { ValidFilters, isValidFilter } from '../../types/valid-filter.type';
@Component({
  selector: 'app-todo-footer',
  standalone: true,
  imports: [CommonModule, MatButtonModule, RouterModule],

  templateUrl: './todo-footer.component.html',
  styleUrls: ['./todo-footer.component.css'],
})
export class TodoFooterComponent implements OnInit {
  filterList: ValidFilters[] = ['all', 'completed', 'pending'];
  currentFilter!: Signal<ValidFilters>;
  itemsLeftList!: Signal<TodoItem[]>;
  todoList!: Signal<TodoItem[]>;

  constructor(private route: ActivatedRoute, private store: Store) {}

  ngOnInit(): void {
    this.currentFilter = this.store.selectSignal(selectFilter);
    this.itemsLeftList = this.store.selectSignal(selectPendingTodos);
    this.todoList = this.store.selectSignal(selectTodoList);

    const filter = this.route.snapshot.fragment;
    if (filter && isValidFilter(filter)) {
      this.store.dispatch(FilterActions.setFilter({ filter }));
    }
  }

  changeFilter(filter: ValidFilters): void {
    if (filter === this.currentFilter()) return;
    this.store.dispatch(FilterActions.setFilter({ filter }));
  }

  clearCompletedTodos(): void {
    this.store.dispatch(TodoActions.clearAllTodos());
  }
}
