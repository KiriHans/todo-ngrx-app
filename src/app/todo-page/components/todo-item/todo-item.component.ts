import { Component, DestroyRef, Input, OnInit, inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import { A11yModule } from '@angular/cdk/a11y';
import { FormControl, ReactiveFormsModule, Validators } from '@angular/forms';
import { SelectInputDirective } from '../../directives/select-input.directive';

/* RXJS Modules */
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';

/* NGRX ModulesÂ¨*/
import { Store } from '@ngrx/store';
import * as TodoActions from '../../store/actions/todo.actions';

/* Angular Material Modules */
import { MatInputModule } from '@angular/material/input';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';

import { TodoItem } from '../../models/todo-item.model';

@Component({
  selector: 'app-todo-item',
  standalone: true,
  imports: [
    CommonModule,
    A11yModule,
    ReactiveFormsModule,
    SelectInputDirective,
    MatInputModule,
    MatCheckboxModule,
    MatButtonModule,
    MatIconModule,
  ],
  templateUrl: './todo-item.component.html',
  styleUrls: ['./todo-item.component.css'],
})
export class TodoItemComponent implements OnInit {
  @Input() todo!: TodoItem;

  private destroyRef = inject(DestroyRef);

  editing = false;
  show = false;

  completed!: FormControl;
  inputTodo!: FormControl;

  constructor(private store: Store) {}

  ngOnInit(): void {
    this.completed = new FormControl(this.todo.completed);
    this.inputTodo = new FormControl(this.todo.text, Validators.required);

    this.completed.valueChanges.pipe(takeUntilDestroyed(this.destroyRef)).subscribe({
      next: () => {
        this.store.dispatch(TodoActions.toggleTodo({ id: this.todo.id }));
      },
    });
  }

  editTodo(): void {
    this.editing = true;
    this.show = true;

    this.inputTodo.setValue(this.todo.text);
  }

  stopEdit(): void {
    this.editing = false;
    this.show = false;

    if (this.inputTodo.invalid || this.inputTodo.value === this.todo.text) return;

    this.store.dispatch(TodoActions.editTodo({ id: this.todo.id, newText: this.inputTodo.value }));
  }

  deleteTodo(): void {
    this.store.dispatch(TodoActions.deleteTodo({ id: this.todo.id }));
  }
}
