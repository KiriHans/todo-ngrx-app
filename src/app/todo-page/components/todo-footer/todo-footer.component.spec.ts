import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TodoFooterComponent } from './todo-footer.component';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { TodoItem } from '../../models/todo-item.model';
import { AppState } from 'src/app/app.reducer';
import { RouterTestingModule } from '@angular/router/testing';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { TitleCasePipe } from '@angular/common';

import { MatButtonHarness } from '@angular/material/button/testing';
import { HarnessLoader } from '@angular/cdk/testing';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';

describe('TodoFooterComponent', () => {
  let component: TodoFooterComponent;
  let fixture: ComponentFixture<TodoFooterComponent>;
  let el: DebugElement;
  let compiled: HTMLElement;

  let titleCasePipe: TitleCasePipe;
  const initialState: AppState = {
    todos: [new TodoItem('test todo')],
    filter: 'all',
  };
  let store: MockStore<AppState>;

  let harnessLoader: HarnessLoader;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [TodoFooterComponent, NoopAnimationsModule, RouterTestingModule],
      providers: [provideMockStore({ initialState })],
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TodoFooterComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    titleCasePipe = new TitleCasePipe();
    store = TestBed.inject(MockStore<AppState>);

    harnessLoader = TestbedHarnessEnvironment.loader(fixture);

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should render a span with a todo-count class', () => {
    const todoCountSpan = el.query(By.css('.todo-count'));
    const count = todoCountSpan.query(By.css('strong')).nativeElement.innerHTML;
    const contentSpan = todoCountSpan.nativeElement.innerHTML;

    expect(todoCountSpan).toBeTruthy();
    expect(count).toBe('1');
    expect(contentSpan).toContain('items left');
  });

  it('should render a list with the corresponding filters', () => {
    const filterList = component.filterList;

    const filterListElement = el.query(By.css('ul.filters'));
    const filterListItems: string[] = filterListElement
      .queryAll(By.css('li'))
      .map((item) => item.nativeElement.textContent);

    expect(filterListElement).toBeTruthy();
    expect(filterListItems).toHaveLength(filterList.length);

    filterListItems.forEach((item, index) => {
      expect(item.trim()).toBe(titleCasePipe.transform(filterList[index]));
    });
  });

  it('should render a button with the text "Clear completed"', () => {
    const button = el.query(By.css('button.clear-completed'));

    expect(button).toBeTruthy();
    expect(button.nativeElement.textContent.trim()).toBe('Clear completed');
  });

  it('should initialize respective signals on OnInit', () => {
    expect(component.currentFilter()).toBe(initialState.filter);
    expect(component.itemsLeftList()).toStrictEqual(initialState.todos);
  });

  it("should change filters when it's respective filter anchor is pressed", async () => {
    const changeFilterSpy = jest.spyOn(component, 'changeFilter');
    const storeDispatchSpy = jest.spyOn(store, 'dispatch');

    const filterList = component.filterList;
    const filterAnchors = await harnessLoader.getAllHarnesses(
      MatButtonHarness.with({ selector: 'a' })
    );

    expect(filterAnchors).toBeTruthy();
    expect(filterAnchors).toHaveLength(filterList.length);

    const [firstFilter, secondFilter, thirdFilter] = filterAnchors;

    await firstFilter.click();

    expect(changeFilterSpy).not.toHaveBeenCalled();

    await secondFilter.click();

    expect(changeFilterSpy).toHaveBeenCalledTimes(1);
    expect(storeDispatchSpy).toHaveBeenCalledTimes(1);

    await thirdFilter.click();

    expect(changeFilterSpy).toHaveBeenCalledTimes(2);
    expect(storeDispatchSpy).toHaveBeenCalledTimes(2);
  });

  it('should clear completed todo elements when "clear completed" button is pressed', async () => {
    const clearCompletedTodosSpy = jest.spyOn(component, 'clearCompletedTodos');
    const storeDispatchSpy = jest.spyOn(store, 'dispatch');

    const cleanCompletedTodosButton = await harnessLoader.getHarness(
      MatButtonHarness.with({ selector: '.clear-completed' })
    );

    await cleanCompletedTodosButton.click();

    expect(clearCompletedTodosSpy).toHaveBeenCalled();
    expect(storeDispatchSpy).toHaveBeenCalled();
  });

  it('should show the correct number of items left in the store', () => {
    const testTodoItems: TodoItem[] = [
      new TodoItem('Item 1'),
      new TodoItem('Item 2'),
      new TodoItem('Item 3'),
    ];
    store.setState({ todos: testTodoItems, filter: 'all' });
    store.refreshState();
    fixture.detectChanges();

    const todoCountSpan = el.query(By.css('.todo-count'));
    const count = todoCountSpan.query(By.css('strong')).nativeElement.innerHTML;

    expect(count).toBe(`${testTodoItems.length}`);
  });
});
