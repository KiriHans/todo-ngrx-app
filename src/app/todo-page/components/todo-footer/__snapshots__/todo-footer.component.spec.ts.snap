// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFooterComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <footer>
    <span
      class="todo-count"
    >
      <strong>
        1
      </strong>
       items left 
    </span>
    <ul
      class="filters"
    >
      <li>
        
        <a
          aria-disabled="false"
          class="mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
          href="/#all"
          mat-button=""
          mat-button-ripple-uninitialized=""
          routerlinkactive="active-filter"
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-button__ripple"
          />
          <span
            class="mdc-button__label"
          >
             All 
          </span>
          <span
            class="mat-mdc-focus-indicator"
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </a>
        
      </li>
      <li>
        <a
          aria-disabled="false"
          class="mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
          href="/#completed"
          mat-button=""
          mat-button-ripple-uninitialized=""
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-button__ripple"
          />
          <span
            class="mdc-button__label"
          >
             Completed 
          </span>
          <span
            class="mat-mdc-focus-indicator"
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </a>
        
        
      </li>
      <li>
        <a
          aria-disabled="false"
          class="mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
          href="/#pending"
          mat-button=""
          mat-button-ripple-uninitialized=""
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-button__ripple"
          />
          <span
            class="mdc-button__label"
          >
             Pending 
          </span>
          <span
            class="mat-mdc-focus-indicator"
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </a>
        
        
      </li>
      
    </ul>
    
    <button
      class="clear-completed mdc-button mdc-button--unelevated mat-mdc-unelevated-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
      fragment="all"
      mat-button-ripple-uninitialized=""
      mat-flat-button=""
      tabindex="0"
    >
      <span
        class="mat-mdc-button-persistent-ripple mdc-button__ripple"
      />
      <span
        class="mdc-button__label"
      >
         Clear completed 
      </span>
      <span
        class="mat-mdc-focus-indicator"
      />
      <span
        class="mat-mdc-button-touch-target"
      />
    </button>
    
  </footer>
</div>
`;
